File: Solutions/117-Populating Next Right Pointers in Each Node II/117-Populating Next Right Pointers in Each Node II.go
//BFS
//https://leetcode.com/problems/populating-next-right-pointers-in-each-node-ii/discuss/961871/golang-BFS-solution
func connect(root *Node) *Node {
    if root == nil {
        return root
    }
    
    q := []*Node{root}
    for len(q) > 0 {
        for i := 0; i < len(q) - 1; i++ {
            q[i].Next = q[i + 1]
        }
        
        tmp := make([]*Node, 0, len(q) * 2)
        for _, n := range q {
            if n.Left != nil {
                tmp = append(tmp, n.Left)
            }
            if n.Right != nil {
                tmp = append(tmp, n.Right)
            }
        }
        q = tmp        
    }
    return root
}

//Another
func connect(root *Node) *Node {
    if root == nil {
        return nil
    }
    queue := make([]*Node, 0)
    queue = append(queue, root)
    for n := len(queue); n > 0; n = len(queue) {
        for i := 0; i < n; i++ {
            curr := queue[0]
            if i != n - 1 {
                curr.Next = queue[1]
            }
            if curr.Left != nil {
                queue = append(queue, curr.Left)
            }
            if curr.Right != nil {
                queue = append(queue, curr.Right)
            }
            queue = queue[1:]
        }
    }
    return root
}

File: Solutions/117-Populating Next Right Pointers in Each Node II/117-Populating Next Right Pointers in Each Node II.py
#BFS
#https://leetcode.com/problems/populating-next-right-pointers-in-each-node-ii/discuss/961834/Python3-simple-BFS-using-deque
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return root
        
        queue = deque([root])
        while queue:
            level = len(queue)                         # count the node in current level
            pre = None
            while level:                               # pop the remain nodes in the same level
                cur = queue.popleft()
                level -= 1
                if cur.left: queue.append(cur.left);
                if cur.right: queue.append(cur.right);
                if pre: pre.next = cur
                pre = cur
            pre.next = None
        return root

#Third fastest
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return
        q=collections.deque([root])
        level=collections.deque([])
        while q:
            # level=[]
            node=q.popleft()
            if node.left:
                level.append(node.left)
            if node.right:
                level.append(node.right)
            
            if q:
                node.next=q[0]
            if not q:
                q,level=level,q
        return root


#Second fastest
from collections import deque
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return root
        Q = deque([root])
        
        while Q:
            size = len(Q)
            
            for i in range(size):
                node = Q.popleft()
                if i < size - 1:
                    node.next = Q[0]
                
                if node.left:
                    Q.append(node.left)
                if node.right:
                    Q.append(node.right)

                node = node.next 
        return root

#Fastest
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if root is None:
            return root
        
        queue = collections.deque()
        queue.append(root)
        while queue:
            length = len(queue)
            pre = None
            for _ in range(length):
                node = queue.popleft()
                if pre is not None:
                    pre.next = node
                pre = node
                if node.left is not None:
                    queue.append(node.left)
                if node.right is not None:
                    queue.append(node.right)
                    
        return root